### PR TITLE
29 implement queue network

### PR DESCRIPTION
**Pull request by Elon-Abulafia**
depends on #67

### CODE REVIEW

File: src/pipert2/core/base/pipe.py [hunk 1]
```diff
@@ -15,30 +17,34 @@ class Pipe:
...
-            networking: Network object responsible for the routine's communication.
+            network: Network object responsible for the routine's communication.
             logger: PipeLogger object for logging the pipe actions.
 
         Attributes:
-            networking: Network object responsible for the routine's communication.
+            network: Network object responsible for the routine's communication.
             logger: PipeLogger object for logging the pipe actions.
             flows (dict[str, Flow]): Dictionary mapping the pipe flows to their name.
             event_board (EventBoard): EventBoard object responsible for the pipe events.
```
Comment (tomerdayan168) [inline src/pipert2/core/base/pipe.py:31]: Description about 'self.default_data_transmitter' is missing

File: src/pipert2/core/base/pipe.py [hunk 2]
```diff
@@ -15,30 +17,34 @@ class Pipe:
...
-    def create_flow(self, flow_name: str, auto_wire: bool, *routines: Routine):
+    def create_flow(self, flow_name: str, auto_wire: bool, *routines: Routine,
+                    data_transmitter: DataTransmitter = None):
         """Create a new flow in the pipe.
 
         Args:
-            flow_name (str): The name of the flow to be created
+            data_transmitter (DataTransmitter): A data transmitter object that indicates how data will be transferred
+                                                inside the flow.
+            flow_name (str): The name of the flow to be created.
```
Comment (tomerdayan168) [inline src/pipert2/core/base/pipe.py:47]: Swap lines between data_transmitter and flow_name by the order of the function's parameters

File: src/pipert2/core/managers/networks/queue_network.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        Args:
+            source: The source routine that generates data.
+            destinations: Destination routines that receive the data.
+            data_transmitter: The data transmitter that indicates how to transfer the data.
+
+        """
+
+        for destination_routine in destinations:
+            queue = Queue(maxsize=1)
```
Comment (tomerdayan168) [inline src/pipert2/core/managers/networks/queue_network.py:44]: Config or TODO config

File: src/pipert2/utils/shared_memory/shared_memory_generator.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    except posix_ipc.ExistentialError:
+        return None
+    except Exception:
+        return None
+
+    mapfile = mmap.mmap(memory.fd, memory.size)
+    memory.close_fd()
+    semaphore.release()
+
+    return SharedMemory(memory, semaphore, mapfile)
```
Comment (tomerdayan168) [inline src/pipert2/utils/shared_memory/shared_memory_generator.py:32]: The standard is - try, except,.., else

File: src/pipert2/core/handlers/message_handler.py [hunk 5]
```diff
@@ -43,18 +46,31 @@ def put(self, message: Message):
...
 
         """
+
         try:
             message = Message.decode(self._get())
+
+            if message is None:
+                message = Message({}, self.routine_name)
+            else:
+                if type(self.receive) == callable:
```
Comment (tomerdayan168) [inline src/pipert2/core/handlers/message_handler.py:69]: else if

File: src/pipert2/core/handlers/message_handlers/queue_handler.py [hunk 6]
```diff
@@ -51,6 +49,9 @@ def _put(self, message: bytes):
 
         """
 
+        if self.output_queue is None:
+            raise Exception(f"{self.routine_name}'s output_queue was not initialized when put was called!")
```
Comment (tomerdayan168) [inline src/pipert2/core/handlers/message_handlers/queue_handler.py:53]: Maybe raise a custom exception instead general exception?

Comment (San-Moshe) [reply src/pipert2/core/handlers/message_handlers/queue_handler.py:53]: +1

File: src/pipert2/core/handlers/message_handlers/queue_handler.py [hunk 7]
```diff
@@ -34,10 +31,11 @@ def _get(self) -> bytes:
...
 
-        try:
-            message = self.input_queue.get(block=self.blocking, timeout=self.timeout)
-        except Empty:
-            print("The queue is empty")  # TODO: Replace with log
+        if self.input_queue is not None:
+            try:
+                message = self.input_queue.get(block=self.blocking, timeout=self.timeout)
+            except Empty:
+                print("The queue is empty")  # TODO: Replace with log
```
Comment (San-Moshe) [inline src/pipert2/core/handlers/message_handlers/queue_handler.py:38]: Replace with log :upside_down_face:

File: tests/pipert/core/base/test_pipe.py [hunk 8]
```diff
@@ -44,9 +44,8 @@ def test_create_flow_auto_wire_true(dummy_pipe: Pipe):
...
 
     for first_routine_mock, second_routine_mock in zip(routine_mocks, routine_mocks[1:]):
-        network_mock.link.assert_any_call(source=first_routine_mock, destinations=second_routine_mock,
-                                          transmit=data_transmitter_mock.transmit(),
-                                          receive=data_transmitter_mock.receive())
+        network_mock.link.assert_any_call(source=first_routine_mock, destinations=(second_routine_mock,),
+                                          data_transmitter=data_transmitter_mock)
 
 
 def test_link(dummy_pipe: Pipe):
```
Comment (San-Moshe) [inline tests/pipert/core/base/test_pipe.py:51]: Add another test that check if correct link were established. 
For example: Routine A MessageHandler Destination queue is the right  destination queue

Comment (San-Moshe) [reply tests/pipert/core/base/test_pipe.py:51]: This is very critical logic in the pipe, we should validate proper behavior precisely.

**Review by San-Moshe — CHANGES_REQUESTED** [review]

File: src/pipert2/core/managers/networks/queue_network.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        """Links between two QueueHandlers of the given routines.
+
+        Args:
+            source: The source routine that generates data.
+            destinations: Destination routines that receive the data.
+            data_transmitter: The data transmitter that indicates how to transfer the data.
+
+        """
+        publish_queue = PublishQueue()
+
```
Comment (MayoG) [inline src/pipert2/core/managers/networks/queue_network.py:44]: I think it won't work for routine that get their data from multiple routines.

for example:     Routine 'B' will get their data from Routines 'A1' and 'A2'.

File: src/pipert2/utils/publish_queue.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        return q
+
+    @ensure_parent
+    def put(self, value, block=False, timeout=1):
+        """Publish a value to every registered queue.
+
+        Args:
+            value: The value to push to every registered queue.
+            block: Whether or not to block the queue when putting a message.
+            timeout: How long to wait if block is true.
```
Comment (MayoG) [inline src/pipert2/utils/publish_queue.py:48]: Technically, the timeout will be the given time multiple by the output queues count.

**Review by MayoG — CHANGES_REQUESTED** [review]

**Review by MayoG — DISMISSED** [review]

**Review by MayoG — APPROVED** [review]

**Review by San-Moshe — APPROVED** [review]

**Pull request merged**